services/lab-service/lab-service/topology.py

```diff
@@ -0,0 +1,247 @@
+A_SW_NAME = A_SW_DEF['name']
+
+
+def pname(sw, port):
+    return '{}-{}'.format(sw, port)
+
+
+def resolve_host(uri):
+    parts = uri.split(":", 3)
+    ip = socket.gethostbyname(parts[1])
+    return parts[0] + ":" + ip + ":" + parts[2]
+
+
+class Switch:
+    def __init__(self, name, of_ver):
+        self.name = name
+        self.of_ver = of_ver
+
+    @classmethod
+    def create(cls, sw_def):
+        # force OF 1.3
+        of_ver = 'OpenFlow13'
+        name = sw_def['name']
+        switch = cls(name, of_ver)
+
+        cmd = [
+            '--if-exists del-br %s' % name,
+            'add-br %s' % name,
+            'set bridge {} other-config:datapath-id={}'.format(name, sw_def['dp_id'].replace(':', '')),
+            'set bridge %s fail_mode=secure' % name,
+            'set bridge {} protocols={}'.format(name, of_ver)
+        ]
+
+        # define ports
+        for pnum in range(1, sw_def['max_port'] + 1):
+            cmd.append('add-port {} {}'.format(name, pname(name, str(pnum))))
+
+        # TODO add vlan filter on tgen ports
+        # out_ports = sw_def.get('out_ports', [])
+        # for port_def in out_ports:
+        #     switch.add_port(pname(name, str(port_def['port'])), str(port_def['port']))
+
+        vsctl(cmd)
+        return switch
+
+    def destroy(self):
+        vsctl(['del-br {}'.format(self.name)])
+
+    def setup_port(self, p_name, p_num, bandwidth=None, batch=False):
+        cmd = [
+            'set interface {} ofport_request={}'.format(p_name, p_num),
+        ]
+        if bandwidth:
+            cmd.append('set interface {} ingress_policing_rate={}'.format(p_name, bandwidth))
+
+        return " -- ".join(cmd) if batch else vsctl(cmd)
+
+    def dump_flows(self):
+        return ofctl(["dump-flows {sw} -O {of_ver}".format(sw=self.name, of_ver=self.of_ver)])[0]
+
+    def mod_port_state(self, ports, port_state):
+        if ports and port_state:
+            ofctl(["mod-port {sw} -O {of_ver} {port} {port_state}"
+                  .format(sw=self.name, of_ver=self.of_ver, port=str(port), port_state=port_state) for port in ports])
+
+    def add_controller(self, controller, batch=False):
+        cmd = [
+            'set-controller {} {}'.format(self.name, controller),
+            'set controller {} connection-mode=out-of-band'.format(self.name)
+        ]
+        return " -- ".join(cmd) if batch else vsctl(cmd)
+
+    def remove_controller(self):
+        return vsctl(['del-controller %s' % self.name])
+
+
+class ASwitch(Switch):
+    def setup(self, a_mappings):
+        self.add_route_flows(a_mappings)
+
+    def add_route_flows(self, mappings):
+        if mappings:
+            ofctl(['add-flow {sw} -O {of_ver} in_port={in_port},actions=output={out_port}'
+                       .format(sw=self.name, of_ver=self.of_ver, **flow) for flow in mappings])
+
+    def del_route_flows(self, mappings):
+        if mappings:
+            ofctl(['del-flows {sw} -O {of_ver} in_port={in_port}'
```

Comment:
We can use cookie(masked or not) to remove all "required" rules/flows.

Reply:
It is ASwith class,  what handle only simple port to port routes. I doesn't see any profit of using addition cookie abstraction on these actions